Introduce lookup runtime fields

dnhatn opened the pull request:
This PR introduces the lookup runtime fields which are used to retrieve data from the related indices. The below search request enriches its search hits with the location of each IP address from the `ip_location` index.

```js
POST logs/_search
{
  "runtime_mappings": {
    "location": {
      "type": "lookup",
      "lookup_index": "ip_location",
      "query_type": "term",
      "query_input_field": "ip",
      "query_target_field": "_id",
      "fetch_fields": [
        "country",
        "city"
      ]
    }
  },
  "fields": [
    "timestamp",
    "message",
    "location"
  ]
}
```

Response:

```js
{
  "hits": {
    "hits": [
      {
        "_index": "logs",
        "_id": "1",
        "fields": {
          "location": [
            {
              "city": [ "Montreal" ],
              "country": [ "Canada" ]
            }
          ],
          "message": [ "the first message" ]
        }
      }
    ]
  }
}
```

dnhatn committed:
Introduce field retrieval joins

dnhatn commented:
run elasticsearch-ci/part-1

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

ywelsch reviewed (commented):
I've just done a quick pass today, and left some questions.

dnhatn commented on docs/reference/search/search-your-data/retrieve-selected-fields.asciidoc:
```diff
@@ -288,6 +288,135 @@ structure of the nested `user` array:
+----
+
+<1> The name of the field. The response of this field is returned under this name.
+<2> The type of the field. Here the `lookup` type means join fields are retrieved using the <<docs-get,GET API>>
```
Good point. I will make this change.

dnhatn committed:
Merge branch 'master' into retrieval-join

dnhatn committed:
Switch to runtime lookup field

dnhatn commented:
run elasticsearch-ci/part-1

dnhatn commented:
@ywelsch @jtibshirani I've updated the PR according to our discussion. Would you mind taking another look? Thank you!

dnhatn committed:
Fix tests

dnhatn committed:
Fix tests

dnhatn commented:
run elasticsearch-ci/part-1

jtibshirani commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -777,6 +813,8 @@ public void executeFetchPhase(ShardFetchRequest request, SearchShardTask task, A
         final ReaderContext readerContext = findReaderContext(request.contextId(), request);
         final ShardSearchRequest shardSearchRequest = readerContext.getShardSearchRequest(request.getShardSearchRequest());
         final Releasable markAsUsed = readerContext.markAsUsed(getKeepAlive(shardSearchRequest));
+        listener = tryFetchLookupFields(shardSearchRequest, task, listener);
```
It would be really nice if we could handle this in the fetch phase (maybe even in `FieldFetcher`) instead of having special handling for it here. Did you choose not to do that because of concerns over blocking the search thread?

dnhatn replied:
That's correct. The FetchPhase is synchronous, while the lookup process is async. I tried to make the FetchPhase async, but it requires more work (mainly because aggregations also use it). I can back out this optimization (i.e., tries to look up using local nodes only) and introduce it later if you and Yannick prefer.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:
```diff
@@ -75,6 +80,10 @@ private static FieldFetcher create(
 
             for (String field : context.getMatchingFieldNames(fieldPattern)) {
                 MappedFieldType ft = context.getFieldType(field);
+                if (ft instanceof LookupRuntimeFieldType lookupField) {
```
I don't have a concrete suggestion yet, but it'd be really nice if we could structure this to avoid type checking. We've worked really hard to keep the mappers modular, so there's no "special casing" in core search code for specific types. In fact many mappers live in modules, so it's not even possible to check their types within search.

dnhatn replied:
Same here. I thought on this before decided to use the type checking. I will play a bit more if I can find a better option than the type checking.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/LookupField.java:
```diff
@@ -0,0 +1,131 @@
+        docFields.put("_id", List.of(lookupId));
+        if (failure != null) {
+            docFields.put(
+                "_failure",
```
I'm unsure about this part of the API where we handle failures. We don't have precedent for it in other parts of the search request, to return failures on a per-document or per-hit basis. It also seems like it could lead to the same failure message being returned several times -- for example if expensive queries were disabled, I think each hit would have the same error message (and the page size could be 100, so that would be 100 copies)?

We've brainstormed ideas for handling fine-grained errors in the context of runtime fields. Maybe for now we could have a really simple approach that matches the current error handling strategy. For most errors we'd error out the whole search. If the lookup document with the right ID was not found, this wouldn't cause the whole search to fail, but it would just be missing from the `fields` output.

dnhatn replied:
I am okay with your suggestion (maybe we can add an option `on_error` for this). Also, if we decide not to return the error, then I think we shouldn't return `_id`.

jtibshirani reviewed (commented):
Thanks @dnhatn for making that big revision. I left a few initial comments focused on how to make the change more modular and fit in the existing search framework.

I also had some overall questions about the API:
* Should we restrict the valid types for `lookup_id` for now, maybe to just `keyword`? It would indicate an error if the person asks for a geopoint for example... now we would just call `toString` on it and not find any IDs.
* Do we have a few examples in mind where loading by ID is sufficient, instead of loading based on another field's value? I'm wondering if it doesn't actually work for the majority of cases, since users don't often set custom IDs.
* Smaller comment, in your example should we be returning `ip` too? It seems redundant with the `_id` field, and the user never actually requested that field, so it could be a little surprising.

dnhatn commented on docs/reference/search/search-your-data/retrieve-selected-fields.asciidoc:
```diff
@@ -288,6 +288,123 @@ structure of the nested `user` array:
 
+[discrete]
+[[lookup-runtime-fields]]
+==== Retrieve fields from related indices
```
++, I will move this to the runtime page.

dnhatn commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchLookupFieldsPhase.java:
```diff
@@ -0,0 +1,227 @@
+            listener.onResponse(null);
+        }
+
+        final MultiGetRequest multiGetRequest = new MultiGetRequest();
```
I used MGET because of its performance: it's faster and uses fewer messages. We can reduce more messages with the optimization that uses local nodes only. We can sync on this.

dnhatn commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -282,6 +285,17 @@ long buildTookInMillis() {
 
     @Override
     protected void doExecute(Task task, SearchRequest searchRequest, ActionListener<SearchResponse> listener) {
+        listener = listener.delegateFailure((innerListener, resp) -> {
```
I had this approach in the previous PR (see https://github.com/elastic/elasticsearch/pull/77097/files#diff-5198131d51dc75f4b8b82d7cc9a3960b0955bbd50c7b6573bdc5a227f02d12cfL52), but it didn't work with the ccsMinimizeRoundtrips enabled. I had to move the lookup process to the end of the search request.

dnhatn commented:
> Should we restrict the valid types for lookup_id for now, maybe to just keyword? It would indicate an error if the person asks for a geopoint for example... now we would just call toString on it and not find any IDs.

I think long and integer types are also valid?

> Do we have a few examples in mind where loading by ID is sufficient, instead of loading based on another field's value? I'm wondering if it doesn't actually work for the majority of cases, since users don't often set custom IDs.

For now, our requirement is lookup using _id. We can ask for the confirmation from Gilad. 

> Smaller comment, in your example should we be returning ip too? It seems redundant with the _id field, and the user never actually requested that field, so it could be a little surprising.

No, the latest version doesn't return that field. I have updated the example.

@jtibshirani Thank you for the review. I think I have responded your points and will address them.

dnhatn commented:
@ywelsch Please hold off your review. I am going to make some major changes in the PR.

dnhatn committed:
lookup by terms

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

elasticsearchmachine commented:
Hi @dnhatn, I've created a changelog YAML for you.

dnhatn committed:
Update docs/changelog/82385.yaml

dnhatn commented:
@ywelsch @jtibshirani I've reworked this PR to support term queries in lookup runtime fields. Would you mind taking another look? Thank you!

ywelsch commented on docs/reference/mapping/runtime.asciidoc:
```diff
@@ -803,6 +808,117 @@ address.
+<4> A field on the searching index whose values are used as the input values of the lookup query
+<5> A field on the lookup index which the lookup query searches against
+<6> A list of fields to retrieve from the lookup index. See the <<search-fields, `fields`>> parameter of a search request.
+<7> An optional parameter to control the maximum number of matching documents from the lookup query. This value must be between 1 and 100.
```
I'm wondering if we should even expose this, or whether this should initially be 1 by default. Once you allow more than one, there's the question of ordering/sorting.

ywelsch reviewed (commented):
Thanks for the update, Nhat. Code looks mostly good, there's probably some API stuff we should discuss (with Julie / Gilad). Perhaps Julie can give this a pass first before we discuss.

dnhatn committed:
wording

dnhatn committed:
remove max_match_size

dnhatn committed:
fix test

dnhatn commented on docs/reference/mapping/runtime.asciidoc:
```diff
@@ -803,6 +808,117 @@ address.
+        "max_match_size": 1 <7>
+    }
+  },
+  "fields": [
```
It supports both wildcards and runtime fields. This is a `fields` parameter of a search request.

dnhatn commented on server/src/main/java/org/elasticsearch/common/document/DocumentField.java:
```diff
@@ -48,16 +50,26 @@ public DocumentField(StreamInput in) throws IOException {
         this.name = Objects.requireNonNull(name, "name must not be null");
         this.values = Objects.requireNonNull(values, "values must not be null");
         this.ignoredValues = Objects.requireNonNull(ignoredValues, "ignoredValues must not be null");
+        this.lookupFields = Objects.requireNonNull(lookupFields, "lookupValues must not be null");
```
thanks, fixed in https://github.com/elastic/elasticsearch/pull/82385/commits/3a0460aadbedce56c6346bc527cfcafb511c6bce

dnhatn commented on docs/reference/search/search-your-data/retrieve-selected-fields.asciidoc:
```diff
@@ -288,6 +288,123 @@ structure of the nested `user` array:
+The response of lookup fields is similar to the response of nested fields.
+Values inside `lookup` fields are grouped to maintain the independence
+of each document from the lookup index. If there is any error during
+the lookup, the error is returned under the `_failure` field.
```
I've removed the `_failure` section to be consistent with other search features as we don't return any document-level failure in search responses.

dnhatn committed:
spotless

dnhatn commented on server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CrossClusterSearchIT.java:
```diff
@@ -236,6 +246,140 @@ public void testCancel() throws Exception {
     }
 
+    /**
+     * Makes sure that lookup fields are resolved on the local cluster
```
Yes, let's discuss this. I chose to resolve lookup fields in only the local cluster to have a consistent result with or without ccs_minimize_roundtrips.

jtibshirani replied:
We discussed this offline and agreed it'd be more natural for the lookup field to be resolved within the same cluster. Related note: I think it would be totally fine if we didn't want to support `ccs_minimize_roundtrips: false` at first. It is not the default and some new features like vector tiles do not support it (https://www.elastic.co/guide/en/elasticsearch/reference/7.17/modules-cross-cluster-search.html#ccs-network-delays).

javanna replied:
I don't have all the context, but it may be good to mention that async search does not support minimizing roundtrips, and that Kibana relies on async search nowadays. It seems like in general we should discuss what to do about minimizing roundtrips.

jtibshirani replied:
This is an important point, then we should indeed support `ccs_minimize_roundtrips: false`. +1 to discuss separately what to do about minimizing roundtrips, it seems it's causing complexity.

dnhatn replied:
Sorry for not being clear. The issue here is not related to ccs_minimize_roundtrips. It's about the user experience (i.e., providing the least surprising behavior for the feature). Also, we don't know if a search request fetches lookup runtime fields defined in the index mappings on the coordinating node where we decide to enable or disable the ccs_minimize_roundtrips.

jtibshirani commented on docs/reference/mapping/runtime.asciidoc:
```diff
@@ -803,6 +808,117 @@ address.
+}
+----
+<1> Define a runtime field in a search request with a type of `lookup` that retrieves fields from the lookup index using the <<search-multi-search,Multi search API>>
+<2> The lookup index where the lookup query executes against
```
For simplicity I think we can avoid allowing wildcard patterns. Other "lookup" functionality like `terms` lookup, get requests, etc., assume you are targeting one index (or index abstraction).

jtibshirani commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -160,6 +162,7 @@ public TransportSearchAction(
         this.namedWriteableRegistry = namedWriteableRegistry;
         this.executorSelector = executorSelector;
         this.defaultPreFilterShardSize = DEFAULT_PRE_FILTER_SHARD_SIZE.get(clusterService.getSettings());
+        this.fetchLookupFieldsPhase = new FetchLookupFieldsPhase(searchTransportService, searchService::isAllowExpensiveQueries);
```
If we changed the CCS behavior so the lookup index is resolved in the same cluster, could we convert this to a "standard" search phase class?

jtibshirani commented on docs/reference/mapping/runtime.asciidoc:
```diff
@@ -803,6 +808,115 @@ address.
+    "location": {
+        "type": "lookup", <1>
+        "index": "ip_location", <2>
+        "query_type": "term", <3>
```
Small comment, it could be nice to exclude "query_type" in this first version to keep things simple. I'm actually not sure how it will look when we extend to other queries, would the user want to specify the query type like this, or use a real query definition plus a search-template like syntax?

Also what do you think of adding some structure similar to this? It could clarify what is associated with the lookup vs. main index.
```
"path": "host",
"lookup": {
    "index": "ip_location",
    "path": "ip",
    "fields": ["country", "city"]
}
```

dnhatn replied:
I found your suggestion is more confusing. I've changed `index` to `lookup_index`.

> it could be nice to exclude "query_type" in this first version to keep things simple.

I can do it, but I think we will introduce the`geo_match` query type soon.

jtibshirani reviewed (commented):
I did another pass, it's looking like great progress to me. I still need to dig into some details (specifically the `LookupFieldCollector` changes).

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:
```diff
@@ -66,6 +67,8 @@ private static FieldFetcher create(
         Map<String, FieldContext> fieldContexts = new LinkedHashMap<>();
         List<String> unmappedFetchPattern = new ArrayList<>();
 
+        Map<String, LookupFieldCollector> lookupFieldCollectors = new HashMap<>();
```
Here's an idea for how to refactor this so we don't need this new `LookupFieldCollector` abstraction:
* Update `ValueFetcher` with a new method `fetchDocumentField` that returns `DocumentField` and can be overridden by `LookupRuntimeFieldType`. Use this new method in `FieldFetcher` instead of `fetchValues`. Include a default implementation that delegates to `fetchValues` so we don't need to update a bunch of classes that already implement `fetchValues`.
* Make `LookupRuntimeFieldType#valueFetcher` be responsible for fetching its own information about the input field values, instead of doing it here in `FieldFetcher`. I think that it could just create its own value fetcher and perform the lookup itself.

I didn't try this out in detail so let me know if I'm missing something 🤔

dnhatn replied:
I tried to avoid looking up input fields twice when they are requested in the fetch fields, but it seems we cache them. I integrated your suggestion in https://github.com/elastic/elasticsearch/pull/82385/commits/bb278b995812e71b55d8bd9abbe47b3f5d51bb14.

jtibshirani reviewed (commented):
I looked through the rest of the PR and left some more comments.

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

dnhatn committed:
Fetch documents instead

dnhatn committed:
Move lookup process to data nodes

dnhatn committed:
spotless

dnhatn commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into retrieval-join

dnhatn committed:
fix assertions

dnhatn committed:
oops

dnhatn commented:
run elasticsearch-ci/part-2

dnhatn commented on server/src/main/java/org/elasticsearch/common/document/DocumentField.java:
```diff
@@ -104,7 +116,16 @@ public void writeTo(StreamOutput out) throws IOException {
+        if (out.getVersion().onOrAfter(Version.V_8_1_0)) {
+            out.writeCollection(lookupFields);
+        } else {
+            // We deliberately do not to fail CCS search requests when the remote clusters are on the new version,
```
We don't have to worry about this anymore as lookup fields are resolved on each cluster now.

dnhatn commented on server/src/internalClusterTest/java/org/elasticsearch/action/search/LookupRuntimeFieldIT.java:
```diff
@@ -0,0 +1,287 @@
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class LookupRuntimeFieldIT extends ESIntegTestCase {
```
We should use `ESIntegTestCase` by default.

dnhatn commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchLookupFieldsPhase.java:
```diff
@@ -0,0 +1,115 @@
+            return;
+        }
+        final MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
+        for (LookupField lookupField : lookupFields) {
```
> Perhaps by grouping the requests into terms queries to cut down on the number of unique search requests?

I considered this before; but it only works with the `_id` field.

dnhatn commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchLookupFieldsPhase.java:
```diff
@@ -0,0 +1,115 @@
+        }
+        if (allowExpensiveQueries.getAsBoolean() == false) {
+            listener.onFailure(
+                new ElasticsearchException(
```
I think we should require `search.allow_expensive_queries` to protect the stability of the cluster. I am happy to discuss this.

dnhatn commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into retrieval-join

dnhatn committed:
fix tests

dnhatn commented:
@ywelsch @jtibshirani I've moved the lookup process to data nodes on each cluster. I tried to make it a sub fetch phase, but it would require more work. We need to make aggregations async first as the top hits aggregations also fetch documents. Would you mind taking another look? Thank you. (BWC test failures are unrelated).

dnhatn committed:
handle failures

dnhatn committed:
Merge branch 'master' into retrieval-join

dnhatn committed:
handle failures

dnhatn committed:
restore SearchService

dnhatn committed:
Merge branch 'master' into retrieval-join

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

dnhatn committed:
wording

dnhatn committed:
Merge branch 'master' into retrieval-join

dnhatn committed:
update BWC version

dnhatn committed:
update BWC version

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

dnhatn committed:
Move lookup to coordinating node

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

dnhatn committed:
remove query_type

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

dnhatn committed:
fix yaml test

dnhatn committed:
wording

dnhatn committed:
correct message

dnhatn commented:
I've updated the PR to move the lookup process to the coordinating node and removed the `query_type` parameter from the lookup runtime field. 

@ywelsch @jtibshirani Can you have another look? Thank you!

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/LookupField.java:
```diff
@@ -0,0 +1,52 @@
+        if (Strings.isEmpty(clusterAlias)) {
+            searchRequest.indices(lookupIndex);
+        } else {
+            searchRequest.indices(clusterAlias + ":" + lookupIndex);
```
Maybe we could use `RemoteClusterAware.buildRemoteIndexName` here. This handles the "local cluster alias" case too, which could remove some special handling?

dnhatn replied:
++, I adjusted the code.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/SearchHit.java:
```diff
@@ -482,6 +483,42 @@ public void setDocumentField(String fieldName, DocumentField field) {
+        final List<String> fields = new ArrayList<>(documentFields.keySet());
+        for (String field : fields) {
+            documentFields.computeIfPresent(field, (k, docField) -> {
+                if (docField.getLookupFields().isEmpty()) {
```
Maybe we don't want to touch a field unless it's a lookup field? I'm wondering a bit about performance here because it's common for hits to have a bunch of fields (Kibana Discover requests all fields `*` by default).

dnhatn replied:
I think the performance is the same, see [HashMap#computeIfPresent](https://github.com/openjdk/jdk/blob/0af356bb4bfee99223d4bd4f8b0001c5f362c150/src/java.base/share/classes/java/util/HashMap.java#L1253).

jtibshirani commented on server/src/main/java/org/elasticsearch/action/search/FetchLookupFieldsPhase.java:
```diff
@@ -0,0 +1,132 @@
+            final String clusterAlias = context.getRequest().getLocalClusterAlias() == null ? cluster.clusterAlias : null;
+            for (LookupField lookupField : cluster.lookupFields) {
+                final SearchRequest searchRequest = lookupField.toSearchRequest(clusterAlias);
+                searchRequest.setCcsMinimizeRoundtrips(context.getRequest().isCcsMinimizeRoundtrips());
```
Maybe it'd be clearer to assert that ccs_minimize_roundtrips is always 'false' if we're hitting any remote cluster. Because if it's 'true', then we would have already handled the lookup in the remote call.

dnhatn replied:
+1, I added the assertion in https://github.com/elastic/elasticsearch/pull/82385/commits/be08cca75326888f596e41e2b882695a78da3257

jtibshirani replied:
Great, but then here we could do just use `false` here to be clear? Because `TransportSearchAction.shouldMinimizeRoundtrips` can return a different response than `context.getRequest().isCcsMinimizeRoundtrips()`.

dnhatn replied:
I think using the ccs_minimize_roundtrips of the original search request is closer to the user's intention. But I am ok with using false. I've adjusted this in https://github.com/elastic/elasticsearch/pull/82385/commits/5f8a3a2c5531a3b0d068ca559f71430896c4d075.

jtibshirani commented on server/src/main/java/org/elasticsearch/common/document/DocumentField.java:
```diff
@@ -104,7 +118,18 @@ public void writeTo(StreamOutput out) throws IOException {
+            out.writeList(lookupFields);
+        } else {
+            if (lookupFields.isEmpty() == false) {
+                assert false : "Lookup fields require all nodes be on 8.2 or later";
```
Any reason we both assert and throw here?

dnhatn replied:
The assertion is to make sure we never call this method in our tests, and tests always abort if we fail to do so. Without the assertion, Elasticsearch might retry and swallow the error.

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/LookupRuntimeFieldType.java:
```diff
@@ -0,0 +1,242 @@
+
+        @Override
+        public List<Object> fetchValues(SourceLookup lookup, List<Object> ignoredValues) throws IOException {
+            assert false : "call #fetchDocumentField() instead";
```
Why do we assert and throw?

jtibshirani reviewed (commented):
Thanks @dnhatn , I think this is getting close. I left a few (mostly small) comments.

Also, we discussed the API offline and agreed to keep the flat structure. I had some suggestions for simplifying the naming, since we try to keep field names pretty concise:
```
"location": {
  "type": "lookup",
  "input_path": "ip",
  "target_index": "ip_location",
  "target_path": "_id",
  "fields": ["country", "city"]
}
```

Or we could do "target_field", "target_index", "input_field", "fetch_fields", either way works for me.

dnhatn committed:
Use RemoteClusterAware.buildRemoteIndexName

elasticsearchmachine commented:
Hi @dnhatn, I've updated the changelog YAML for you.

dnhatn committed:
Update docs/changelog/82385.yaml

dnhatn committed:
Rename fields

dnhatn committed:
Assert ccs_minimize_roundtrips

dnhatn committed:
add size for future optimization

dnhatn committed:
wording

dnhatn committed:
Merge remote-tracking branch 'dnhatn/retrieval-join' into retrieval-join

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

dnhatn committed:
fix tests

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

dnhatn committed:
fix tests

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

dnhatn commented on docs/reference/mapping/runtime.asciidoc:
```diff
@@ -803,6 +808,115 @@ address.
+----
+// TESTRESPONSE[s/"took": 3/"took": $body.took/]
+
+The response of lookup fields are grouped to maintain the independence
```
I've updated the description. We expect a lookup query for a single input value matches at most one document.

dnhatn commented on docs/reference/mapping/runtime.asciidoc:
```diff
@@ -803,6 +808,115 @@ address.
 
+
+[[lookup-runtime-fields]]
+==== Retrieve fields from related indices
```
We can discuss this with Gilad.

dnhatn commented on server/src/main/java/org/elasticsearch/index/mapper/LookupRuntimeFieldType.java:
```diff
@@ -0,0 +1,242 @@
+                return null;
+            }
+            final List<LookupField> lookupFields = inputDoc.getValues().stream().map(input -> {
+                final TermQueryBuilder query = new TermQueryBuilder(queryTargetField, input.toString());
```
> I'm wondering why we don't use a terms query (plural)? Why do we make a separate query for each value?

Yes, this is possible. I can push some optimizations including this suggestion in follow-up works.

> If I'm understanding correctly, then right now we have these surprising semantics. Say the query document has "input_field: [value1, value2]" and the target document has "target_field: [value1, value2]". Then the target document will actually show up twice in the list of fetched documents.

Unless both `input1` and `input2` matches documents that have the same values for the fetch fields. I don't think we should deduplicate the enriched fields.

jtibshirani replied:
Okay, optimizing this in a follow-up sounds good.

dnhatn commented:
> Or we could do "target_field", "target_index", "input_field", "fetch_fields", either way works for me.

I've updated the API using these fields (see https://github.com/elastic/elasticsearch/pull/82385/commits/d33b0d3baf025dc38b7afbcbe590c1bf81e7c189). Thank you for your reviews and discussion. I think this is ready again. Would you mind taking another look?

jtibshirani reviewed (commented):
Thanks @dnhatn, this looks good to me. Before merging we can discuss these points together:
* Whether to release this as experimental
* Double-check the behavior around having multiple input values or multiple matches per value

dnhatn committed:
Set ccsMinimizeRoundtrips to false

dnhatn committed:
Move test to LookupRuntimeFieldTypeTests

dnhatn commented on server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java:
```diff
@@ -1046,6 +1048,105 @@ public void testTooManyUnmappedFieldWildcardPattern() throws IOException {
         expectThrows(TooComplexToDeterminizeException.class, () -> fetchFields(mapperService, source, fieldAndFormatList));
     }
 
+    public void testCollectLookupFields() throws IOException {
```
+1, I moved this test and added some more in https://github.com/elastic/elasticsearch/pull/82385/commits/31632ca84f738a7d681275a120a998c7da211538.

dnhatn committed:
experimental feature

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into retrieval-join

jtibshirani reviewed (approved):
This looks good to me. Since we are releasing it as "experimental", it gives room to tweak the behavior around multiple values, multiple matching docs, etc., so I don't feel we need to iron that out in detail before merging.

dnhatn commented:
@jtibshirani @ywelsch Thank you so much for your thoughtful reviews and feedback on this PR. I really appreciate it.

dnhatn commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into retrieval-join
